Cabal/tests/PackageTests/PackageTester.hs

```diff
@@ -1,367 +1,540 @@
+-- subtests.)  To clean, you ONLY need to delete this directory.
+topDir :: TestM FilePath
+topDir = do
+    test_dir <- testDir
+    (_, test) <- ask
+    return $ test_dir </>
+                case testSubName test of
+                    Nothing -> "dist-test"
+                    Just n -> "dist-test." ++ n
+
+prefixDir :: TestM FilePath
+prefixDir = do
+    top_dir <- topDir
+    return $ top_dir </> "usr"
+
+-- | The absolute path to the build directory that should be used
+-- for the current package in a test.
+distDir :: TestM FilePath
+distDir = do
+    top_dir <- topDir
+    (_, test) <- ask
+    return $ top_dir </> testCurrentPackage test </> "dist"
+
+definitelyMakeRelative :: FilePath -> FilePath -> FilePath
+definitelyMakeRelative base0 path0 =
+    let go [] path = joinPath path
+        go base [] = joinPath (replicate (length base) "..")
+        go (".":xs) ys = go xs ys
+        go xs (".":ys) = go xs ys
+        go (x:xs) (y:ys)
+            | x == y    = go xs ys
+            | otherwise = go (x:xs) [] </> go [] (y:ys)
+    in go (splitPath base0) (splitPath path0)
+
+-- hpc is stupid and doesn't understand absolute paths.
+relativeDistDir :: TestM FilePath
+relativeDistDir = do
+    dist_dir0 <- distDir
+    pkg_dir <- packageDir
+    return $ definitelyMakeRelative pkg_dir dist_dir0
+
+-- | The absolute path to the shared package database that should
+-- be used by all packages in this test.
+sharedDBPath :: TestM FilePath
+sharedDBPath = do
+    top_dir <- topDir
+    return $ top_dir </> "packagedb"
 
-cabal_configure :: SuiteConfig -> PackageSpec -> IO Result
-cabal_configure config spec = do
-    res <- doCabalConfigure config spec
-    record spec res
-    return res
-
-doCabalConfigure :: SuiteConfig -> PackageSpec -> IO Result
-doCabalConfigure config spec = do
-    cleanResult@(_, _, _) <- cabal config spec [] ["clean"]
-    requireSuccess cleanResult
-    res <- cabal config spec []
-           -- Use the package dbs from when we configured cabal rather than any
-           -- defaults.
-           (["configure", "--user", "-w", ghcPath config, "--package-db=clear"]
-            ++ packageDBParams (packageDBStack config)
-            ++ configOpts spec)
-    return $ recordRun res ConfigureSuccess nullResult
+------------------------------------------------------------------------
+-- * Running cabal
+
+cabal :: String -> [String] -> TestM Result
+cabal cmd extraArgs0 = do
+    (suite, test) <- ask
+    prefix_dir <- prefixDir
+    when ((cmd == "register" || cmd == "copy") && not (testPackageDb test)) $
+        error "Cannot register/copy without using 'withPackageDb'"
+    let extraArgs1 = case cmd of
+            "configure" ->
+                -- If the package database is empty, setting --global
+                -- here will make us error loudly if we try to install
+                -- into a bad place.
+                [ "--global"
+                -- TODO: Should be configurable, but it's tricky:
```

Comment:
ghcPath and ghcPkgPath always come from Cabal's LBI, so in that sense it's not configurable. I've moved this comment somewhere more reasonable.